sky/utils/controller_utils.py

```diff
@@ -473,20 +473,37 @@ def get_controller_resources(
     # the clouds of the resources if the controller does not exist.
     # TODO(tian): Consider respecting the regions/zones specified for the
     # resources as well.
-    requested_clouds: Set['clouds.Cloud'] = set()
+
+    requested_clouds_with_region_zone: Dict[str, Dict[str, Set[str]]] = {}
     for resource in task_resources:
-        # cloud is an object and will not be able to be distinguished by set.
-        # Here we manually check if the cloud is in the set.
-        if resource.cloud is not None:
-            if not clouds.cloud_in_iterable(resource.cloud, requested_clouds):
+        cloud_name = str(resource.cloud) if resource.cloud is not None else None
+        if cloud_name is not None:
+            if cloud_name not in requested_clouds_with_region_zone:
                 try:
                     resource.cloud.check_features_are_supported(
                         resources.Resources(),
                         {clouds.CloudImplementationFeatures.HOST_CONTROLLERS})
                 except exceptions.NotSupportedError:
                     # Skip the cloud if it does not support hosting controllers.
                     continue
-                requested_clouds.add(resource.cloud)
+                requested_clouds_with_region_zone[cloud_name] = {}
+            if resource.region is None:
+                requested_clouds_with_region_zone[cloud_name] = {
+                    '_allow_any_region': {'_allow_any_zone'}
+                }
```

Comment:
Yes, it is possible, but it requires special handling to determine whether an empty set indicates that any region is allowed, which means we should only add new regions if it’s the first time. Using a placeholder makes it easier. Maybe `None` could be used as a placeholder instead of using specific strings? It seems that if any region/zone is okay, the fields should be set to `None`.